optuna/progress_bar.py

```diff
@@ -59,22 +76,35 @@ def _init_valid(self) -> None:
         optuna_logging.disable_default_handler()
         optuna_logging._get_library_root_logger().addHandler(_tqdm_handler)
 
-    def update(self, elapsed_seconds: Optional[float]) -> None:
+    def update(self, elapsed_seconds: float) -> None:
         """Update the progress bars if ``is_valid`` is :obj:`True`.
 
         Args:
             elapsed_seconds:
                 The time past since :func:`~optuna.study.Study.optimize` started.
         """
+
         if self._is_valid:
-            self._progress_bar.update(1)
-            if self._timeout is not None and elapsed_seconds is not None:
-                self._progress_bar.set_postfix_str(
-                    "{:.02f}/{} seconds".format(elapsed_seconds, self._timeout)
-                )
+            if self._n_trials is not None:
+                self._progress_bar.update(1)
+                if self._timeout is not None:
+                    self._progress_bar.set_postfix_str(
+                        "{:.02f}/{} seconds".format(elapsed_seconds, self._timeout)
+                    )
+
+            else:
+                assert self._timeout is not None
```

Comment:
We can call `study.optimize` without setting neither `n_trials` nor `timeout` as follows.
```
study.optimize(objective, callbacks=[optuna.study.MaxTrialsCallback(10)])
```
I think we should divide the cases as follows.
- `n_trials` is not None
- `timeout` is not None
- Otherwise (A progress bar cannot be enabled.)

Reply:
Testing with callback you've provided, behavior is as expected - progress bar is disabled even with `show_progress_bar=True`. We are disabling progbar when neither `n_trial` nor `timeout` are specified by ensuring `is_valid=False` in constructor. This `assert` is there only to make mypy happy, and should never be actually raised. I see how this might be not the most readable solution. Alternatively, i could refactor into

```python
if self._n_trials is not None:
  [..]
elif self._timeout is not None:
  [..]
else:
  raise RuntimeError("Should not reach.")
```

I've seen similar codepath in optuna somewhere.

Reply:
Thanks you for your explanation. I missed the flag of `self._is_valid`. It looks like easier to read as you suggested. Could you fix them?
```python
if self._n_trials is not None:
  [..]
elif self._timeout is not None:
  [..]
else:
  assert False
```

Reply:
Done in 620c39be9b9bc8b93c0df11a1466a51a06b91aac!